actionpack/test/controller/parameters/parameters_permit_test.rb

```diff
@@ -377,6 +377,10 @@ def walk_permitted(params)
     assert_equal "32", @params[:person].permit([ :age ])[:age]
   end
 
+  test "to_s falls back to Hash#to_s" do
+    assert_equal @params.to_unsafe_h.to_s, @params.to_s
```

Comment:
Prefer to use the string as assertion here since the implementation may change in the future to not reflect `to_usafe_h` anymore.

```diff
--- a/actionpack/test/controller/parameters/accessors_test.rb
+++ b/actionpack/test/controller/parameters/accessors_test.rb
@@ -35,6 +35,11 @@ class ParametersAccessorsTest < ActiveSupport::TestCase
     assert @params.as_json.key? "person"
   end

+  test "to_s returns the string representation of the parameters hash" do
+    assert_equal '{"person"=>{"age"=>"32", "name"=>{"first"=>"David", "last"=>"Heinemeier Hansson"}, ' \
+      '"addresses"=>[{"city"=>"Chicago", "state"=>"Illinois"}]}}', @params.to_s
+  end
+
   test "each carries permitted status" do
     @params.permit!
     @params.each { |key, value| assert(value.permitted?) if key == "person" }
```